### PR TITLE
PyTorch Lightning GAN tutorial

### LINKED ISSUE #268: PyTorch Lightning example + OpenFL
Integrate example of [PyTorch Lightning](https://www.pytorchlightning.ai) use-case in Federated Learning setup using OpenFL.

### PR DESCRIPTION
**Pull request by mansishr**
Generative Adversarial Networks (GAN) tutorial using PyTorch Lightning on MNIST dataset.
Fixes https://github.com/intel/openfl/issues/268

### CODE REVIEW

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+Cuurently, the tutorial shows how to train both the generator and the discriminator models parallely. Individual models can be trained as well. To train only the generator, the flag 'train_gen_only' should be set to 1 and to train only the discriminator, 'train_disc_only' should be set to 1.
+
+### 5. Links
+* [Original paper](https://arxiv.org/abs/1406.2661)
+* [Original Github code](https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/basic-gan.html)
+
+### 6. How to run this tutorial (without TLS and locally as a simulation):
+
+Go to example folder:
+cd <openfl_folder>/openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md]: Maybe make sense to create env variable with root tutorial path in order to allow user copy/paste parst with `run director` and `run envoy`.
Proposal:
```bash
export PYTORCH_LIGHTNING_MNIST_GAN=<openfl_folder>/openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN

cd $PYTORCH_LIGHTNING_MNIST_GAN/director

cd $PYTORCH_LIGHTNING_MNIST_GAN/envoy

cd $PYTORCH_LIGHTNING_MNIST_GAN/workspace

```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/envoy_config_no_gpu.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+params:
+  cuda_devices: []
+
+optional_plugin_components: {}
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/envoy_config_no_gpu.yaml]: Is this line required? If not, maybe we need to remove it

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/sd_requirements.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+numpy
+pillow
+pytorch-lightning>=1.3
+torch==1.9.1
+torchvision==0.10.1
+torchmetrics>=0.3
+tensorflow==2.7.0
+pynvml==11.4.1
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/sd_requirements.txt]: Would be greate put it after `pytorch-lightning>=1.3` :wink:

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/plugin_for_multiple_optimizers.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        super().__init__()
+
+    @staticmethod
+    def get_tensor_dict(model, optimizer=None):
+        """
+        Extract tensor dict from a model and a list of optimizers.
+
+        Returns:
+        dict {weight name: numpy ndarray}
+        """
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/plugin_for_multiple_optimizers.py]: If it list of optimizers, maybe we need to name it `optimizers`?

**Review by dmitryagapov — COMMENTED** [review]
Also would be great to define typings for functions params and for return values :wink:

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+### 2. About framework: PyTorch Lightning
+[Pytorch Lightning](https://www.pytorchlightning.ai/) is a framework built on top of PyTorch that allows the models to be scaled without the boilerplate.
+
+### 3. About dataset: MNIST
+[MNIST](http://yann.lecun.com/exdb/mnist/) database is a database of handwritten digits that has a training set of 60,000 examples, and a test set of 10,000 examples. It is a subset of a larger set available from NIST. The digits have been size-normalized and centered in a fixed-size image.
+
+### 4. Using multiple optimizers
+The example uses two different optimizers: one for discriminator and one for generator. The plugin to support multiple optimizers with OpenFL has been added (workspace/plugin_for_multiple_optimizers.py). Note that in order to use PyTorch Lightning framework with a single optimizer, this plugin is not required.
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md]: Just simple suggestion: 
```suggestion
The example uses two different optimizers: one for discriminator and one for generator. The [plugin](workspace/plugin_for_multiple_optimizers.py) to support multiple optimizers with OpenFL has been added. Note that in order to use PyTorch Lightning framework with a single optimizer, this plugin is not required.
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+### 5. Training Generator and Discriminator models separately
+Cuurently, the tutorial shows how to train both the generator and the discriminator models parallely. Individual models can be trained as well. To train only the generator, the flag 'train_gen_only' should be set to 1 and to train only the discriminator, 'train_disc_only' should be set to 1.
+
+### 5. Links
+* [Original paper](https://arxiv.org/abs/1406.2661)
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md]:
```suggestion
* [Original GAN paper](https://arxiv.org/abs/1406.2661)
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+* [Original Github code](https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/basic-gan.html)
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md]: Why it called "Original Github code"? link does not lead to github =) Suggestion;
```suggestion
* [Original PyTorch-Lightning training code](https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/basic-gan.html)
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+2. Run envoy:
+```sh
+cd envoy
+bash start_envoy.sh
+```
+
+Optional: start second envoy:
+ - Copy `envoy` to another folder, change the config and run from there:
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md]: I think, we have to change not only config, but -n parameter inside `start_envoy.sh` script as well to start second envoy.

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+### 5. Training Generator and Discriminator models separately
+Cuurently, the tutorial shows how to train both the generator and the discriminator models parallely. Individual models can be trained as well. To train only the generator, the flag 'train_gen_only' should be set to 1 and to train only the discriminator, 'train_disc_only' should be set to 1.
+
+### 5. Links
+* [Original paper](https://arxiv.org/abs/1406.2661)
+* [Original Github code](https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/basic-gan.html)
+
+### 6. How to run this tutorial (without TLS and locally as a simulation):
+
+Go to example folder:
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md]: Suggestion:
```suggestion
Go to example [folder](./)
```
and remove next line:
`cd <openfl_folder>/openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN`

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 10]
```diff
@@ -0,0 +1,759 @@
+{
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb:1]: Could you please try to format all jupyter code by `Jupyterlab Code Formatter`? [Here](https://jupyterlab-code-formatter.readthedocs.io/en/latest/) is a short steps to install this tool. This tool availible only in jupyter-lab, so `JupyterLab Code Formatter` adds a button to Jupyter Lab to automatically format the code in notebooks. 


I guess below are the commands to be done according to the  `Jupyterlab Code Formatter` documentation:
```console
pip install jupyterlab
pip install jupyterlab_code_formatter
pip install black isort
```
Then, restart jupyter-Lab and use special button to automatically format the code in `PyTorch_Lightning_GAN.ipynb` notebook

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]: Is this the current year? @alexey-gruzdev Do we need to update copyright section here?

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        self.data_by_type = {
+            'train': (x_train, y_train),
+            'val': (x_val, y_val)
+        }
+
+    def get_shard_dataset_types(self) -> List[str]:
+        """Get available shard dataset types."""
+        return list(self.data_by_type)
+
+    def get_dataset(self, dataset_type='train'):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]:
```suggestion
    def get_dataset(self, dataset_type='train') -> MnistShardDataset:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    ):
+        """Initialize MnistShardDescriptor."""
+        self.rank, self.worldsize = tuple(int(num) for num in rank_worldsize.split(','))
+        (x_train, y_train), (x_val, y_val) = self.download_data()
+        self.data_by_type = {
+            'train': (x_train, y_train),
+            'val': (x_val, y_val)
+        }
+
+    def get_shard_dataset_types(self) -> List[str]:
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]: I think returned type annotation should be different
```suggestion
    def get_shard_dataset_types(self) -> List[Dict[str, Any]]:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+            raise Exception(f'Wrong dataset type: {dataset_type}')
+        return MnistShardDataset(
+            *self.data_by_type[dataset_type],
+            data_type=dataset_type,
+            rank=self.rank,
+            worldsize=self.worldsize
+        )
+
+    @property
+    def sample_shape(self):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]:
```suggestion
    def sample_shape(self) -> List[str]:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        """Return the sample shape info."""
+        return ['28', '28']
+
+    @property
+    def target_shape(self):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]:
```suggestion
    def target_shape(self) -> List[str]:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+from openfl.interface.interactive_api.shard_descriptor import ShardDataset
+from openfl.interface.interactive_api.shard_descriptor import ShardDescriptor
+
+logger = logging.getLogger(__name__)
+
+
+class MnistShardDataset(ShardDataset):
+    """Mnist Shard dataset class."""
+
+    def __init__(self, x, y, data_type, rank=1, worldsize=1):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]:
```suggestion
    def __init__(self, x, y, data_type, rank: int = 1, worldsize: int = 1) -> None:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+
+class MnistShardDescriptor(ShardDescriptor):
+    """Mnist Shard descriptor class."""
+
+    def __init__(
+            self,
+            rank_worldsize: str = '1, 1',
+            **kwargs
+    ):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]:
```suggestion
    ) -> None:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        self.worldsize = worldsize
+
+        self.x = x[self.rank - 1::self.worldsize]
+        self.y = y[self.rank - 1::self.worldsize]
+
+    def __getitem__(self, index: int):
+        """Return an item by the index."""
+        return self.x[index], self.y[index]
+
+    def __len__(self):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py]:
```suggestion
    def __len__(self) -> int:
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb (hunk 10 above)
Comment (mansishr) [reply openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb:1]: thanks for recommending this @alexey-khorkin

**Comment by mansishr** [comment]
Jenkins please retry a build.

**Comment by mansishr** [comment]
Jenkins please retry a build

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 19]
```diff
@@ -0,0 +1,789 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f821972",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "imgs = fed_dataset.get_valid_data()\n",
+    "\n",
+    "val_imgs = []\n",
+    "for x in range(0, 10):\n",
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: Range up to 10 but we see 6 images, It's a bit non-intuitive. So I guess we can remove range by this:
```python3
val_imgs = fed_dataset.get_valid_data()
```

**Comment by igor-davidyuk** [comment]
Mansi, could you please clean up the notebook, just remove unused, commented, and redundant code.

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 20]
```diff
@@ -0,0 +1,771 @@
...
+   "source": [
+    "TI = TaskInterface()\n",
+    "\n",
+    "import tqdm\n",
+    "\n",
+    "\n",
+    "# The Interactive API supports registering functions definied in main module or imported.\n",
+    "def function_defined_in_notebook(some_parameter):\n",
+    "    print(f\"Also I accept a parameter and it is {some_parameter}\")\n",
+    "\n",
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]:
```suggestion

```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 21]
```diff
@@ -0,0 +1,771 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f821972",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "val_imgs = []\n",
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: Can we just use something like this:
```suggestion
    "val_imgs, _ = next(iter(fed_dataset.get_valid_loader())) # take images from batch \n",
    "z = torch.randn(val_imgs.shape[0], 100) \n"
```
to reuse our code and write it less?

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 22]
```diff
@@ -0,0 +1,771 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9549d1ab",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "last_model.train_generator(val_imgs, z, logdir=\"generated_images\")"
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: Can we add pass command at the end of this cell to remove useless cell output? Like:
```suggestion
    "last_model.train_generator(val_imgs, z, logdir=\"generated_images\")\n",
    "pass"
```

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb (hunk 20 above)
Comment (mansishr) [reply openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: @igor-davidyuk what needs to be changed here? Should I remove the comment?

**Review by alexey-khorkin — APPROVED** [review]
Looks good to me!

**Review by dmitryagapov — APPROVED** [review]

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+export PYTORCH_LIGHTNING_MNIST_GAN=<openfl_folder>/openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN
+```
+
+1. Run director:
+```sh
+cd $PYTORCH_LIGHTNING_MNIST_GAN/director
+bash start_director.sh
+```
+
+2. Run envoy:
```
Comment (alexey-gruzdev) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/README.md:35]: Are we missing here `pip install -r sd_requirements.txt` step?

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/sd_requirements.txt [hunk 24]
```diff
@@ -0,0 +1,8 @@
+numpy
+pillow
+pytorch-lightning>=1.3
+pynvml==11.4.1
+tensorflow==2.7.0
```
Comment (alexey-gruzdev) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/sd_requirements.txt]: Do we really need TensorFlow/PyTorch Lightning/Torchmetrics in ShardDescriptor requirements?

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 25]
```diff
@@ -0,0 +1,744 @@
...
+    "            return self.train_discriminator(imgs, z)\n",
+    "\n",
+    "    def train_generator(self, imgs, z, logdir=\"../generated_images\"):\n",
+    "        self.generated_imgs = self(z)\n",
+    "        sample_imgs = self.generated_imgs[:6]\n",
+    "        sample_imgs = np.reshape(sample_imgs.detach().cpu().numpy(), (6, 28, 28, 1))\n",
+    "\n",
+    "        if not os.path.exists(logdir):\n",
+    "            os.mkdir(logdir)\n",
+    "        tf.autograph.set_verbosity(0)\n",
```
Comment (alexey-gruzdev) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: Can we avoid usage of TensorFlow at all? Looks like we are downloading TensorFlow for only 3 pieces of code :
```
tf.autograph.set_verbosity(0)
        file_writer = tf.summary.create_file_writer(logdir)

        with file_writer.as_default():
            tf.summary.image(
                "Generated images",
                np.reshape(sample_imgs, (-1, 28, 28, 1)),
                max_outputs=10,
                step=0,
            )
```
I am pretty sure we can avoid doing that.

**Comment by alexey-gruzdev** [comment]
Let's also remove `start_envoy_with_tls.sh`, `start_director_with_tls.sh`, since they are just copied, and there is no reference from README.md

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb (hunk 20 above)
Comment (alexey-gruzdev) [reply openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: Let's remove the function itself as @igor-davidyuk  suggested

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb [hunk 26]
```diff
@@ -0,0 +1,744 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f5c09f8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# View it in tensorboard\n",
+    "\n",
+    "# %load_ext tensorboard\n",
```
Comment (alexey-gruzdev) [inline openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: I believe we only need 1 way to launch TensorBoard, please double-check and remove one of this

File: openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb (hunk 25 above)
Comment (mansishr) [reply openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb]: @alexey-gruzdev we are using tensorboard to show generated images. Do we want to view images directly in the notebook?

**Pull request merged**